sampletester/parser.py

```diff
@@ -0,0 +1,150 @@
+
+    This function parses `content` and extracts all YAML documents it finds
+    there.  The documents are stored in lists of `Document` objects grouped and
+    indexed by the value of their top-level 'type' field.  If `strict` was set,
+    untyped documents raise an exception. Otherwise, if a resolver was provided,
+    it is called for each untyped document and the document is reclassified
+    according the type value returned by the provider. If no resolved was
+    provided, the untyped documents are put into their own list with type given
+    by `SCHEMA_TYPE_ABSENT`.
+    """
+    self.add_documents(*[Document(file_name, doc) for doc in yaml.load_all(content)])
+
+  def add_documents(self, *documents: Document):
+    """Adds each doc in `documents` under the right schema type key."""
+    for doc in documents:
+      specified_type = (doc.obj.get(SCHEMA_TYPE_KEY, None)
+                        if isinstance(doc.obj,  dict)
+                        else None)
+
+      if not specified_type:
+        msg = f'no top-level "{SCHEMA_TYPE_KEY}" field specified'
+        if self.strict:
+          raise SyntaxError(msg)
+        logging.warning(msg)
+
+      if not isinstance(specified_type, str):
+        msg = (f'top level "{SCHEMA_TYPE_KEY}" field is not '
+               f'a string: {specified_type}')
+        if self.strict:
+          raise SyntaxError(msg)
+        logging.warning(msg)
+        specified_type = SCHEMA_TYPE_ABSENT
+
+      type_name = specified_type.split(SCHEMA_TYPE_SEPARATOR, 1)[0]
+      self._add_one(type_name, doc)
+    self.resolve_uncategorized()
+
+
+  def _add_one(self, type_name: str, doc: Document):
+    """Adds `doc` to the list of documents with the given type."""
+    similar_docs = self.keyed_docs.get(type_name, [])
+    if not similar_docs:
+      self.keyed_docs[type_name] = similar_docs
+    similar_docs.append(doc)
+
+  def of_type(self, type_name: str) -> List[Document]:
+    """Returns a list of all `Document`s with the given type."""
+    return self.keyed_docs.get(type_name, [])
+
+  def resolve_uncategorized(self):
+    """Categorizes all documents of unknown type by calling the resolver."""
+    if not self.resolver:
+      return
+
+    unknowns = self.of_type(SCHEMA_TYPE_ABSENT)
+    for idx, unknown_doc in enumerate(unknowns):
+      new_type = self.resolver(unknown_doc)
+      if not new_type or new_type == SCHEMA_TYPE_ABSENT:
+        continue
+
+      # mark the doc as now being of the new type, and no longer of unknown type
+      self._add_one(new_type, unknown_doc)
+      unknowns[idx]=None
+
+    # remove `None`s from the list of unknown docs
+    self.keyed_docs[SCHEMA_TYPE_ABSENT] = [doc for doc in unknowns if doc]
+
+
+def only_files_in(paths: Set[str]) -> Set[str]:
+  """Returns only those elements of `paths` that are files"""
+  return {fname for fname in paths if os.path.isfile(fname)}
+
+
+class SyntaxError(Exception):
```

Comment:
This is going to clobber the builtin `SyntaxError`. Why not just use that?

Reply:
Might be good to pick a different name then so that they are unambiguously different exceptions.

Reply:
OK, did that.